### PR TITLE
Decoder for some C instructions 

### PR DESCRIPTION
**Pull request by AyoubJalali**
Hello @silabs-robin @MikeOpenHWGroup, so this PR is related to what we discuss last week, creating an independent decoder from the disassemble of spike, there is nothing smart there just simple systemverilog functions

Please have a look and let's discuss

### CODE REVIEW

File: lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv
Comment (silabs-robin) [inline lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv:508]: The bugs we had before came to existence precisely because people didn't understand what the signal names meant, and then things got connected wrong.

Therefore, I think it is worth spending a small effort to be smart about the naming scheme here.

@MikeOpenHWGroup and @AyoubJalali do you agree it is worth the effort or should we just be pragmatic and move forward quickly?
I'll accept your votes and align with whatever the majority decides.

--------

My proposal:

1. Prefix `isadecode_*`.
    * So everything decoder-related can be easily grouped and identified.
2. Consistently use the same signal names.
    * E.g. `abc_xyz = prefix_abc_xyz` (instead of splitting it like `abc_xyz = abc_prefix_xyz`).
3. Consistenly use exactly the same signal names.
    * E.g. `c_rdp = prefix_c_rdp` (instead of tiny differences like `c_rdp = prefix_c_rd`, without the "p").
4. Keep a separation between "raw immediate" vs "interpreted immediate", etc [0]
    * E.g. `blabla_raw` and `blabla_interpreted` (because toggle coverage of `{imm[20:1], 1'b 0}` will never be reached)

Conclusion:
`c_decode_rd` should be `isadecode_c_rdp_interpreted`, i.e. `prefix_signalname_immediatelevel`, or something similarly orderly.

--------

[0]
There are multiple levels here.
ISACOV has uses for number 3 and 4, AFAIK.
1. Raw binary word - `imm[20|10:1|11|19:12] | rd | opcode`
2. Raw bit field - `imm[20|10:1|11|19:12]`
3. Sorted bit field - `imm[20] | imm[19:12] | imm[11] | imm[10:1]`
4. Interpreted immediate - `imm[20] | imm[19:12] | imm[11] | imm[10:1] | 1'b 0`  (appending 0, or +8, etc)
5. Applied immediates - E.g. `interpreted_immediate + rs1_value`

File: lib/uvm_agents/uvma_isacov/uvma_isacov_mon.sv
Comment (silabs-robin) [inline lib/uvm_agents/uvma_isacov/uvma_isacov_mon.sv:170]: This looks much better. Nice to get rid of some of that spike dasm confusion.

**Review by silabs-robin — CHANGES_REQUESTED** [review]

File: lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv
Comment (AyoubJalali) [reply lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv:508]: > @MikeOpenHWGroup and @AyoubJalali do you agree it is worth the effort or should we just be pragmatic and move forward quickly?

I agree, the challenge here i think is in the naming more than development of methods , but give some time to think more about the naming i don't wanna naming too long specially for signals (for method that's all right I vote for your suggestion)

Comment (MikeOpenHWGroup) [reply lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv:508]: > @MikeOpenHWGroup and @AyoubJalali do you agree it is worth the effort or should we just be pragmatic and move forward quickly?

My short answer is "yes, it is worth the effort".  I also agree that we should focus on the names of methods to make it clear how to use the new decoder.

Comment (MikeOpenHWGroup) [reply lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv]: @silabs-robin proposals:
> 1. Prefix isadecode_*.
So everything decoder-related can be easily grouped and identified.

AFAICT, @AyoubJalali has used `decode_*`.  This is a reasonable comprise.

> 2. Consistently use the same signal names.
E.g. abc_xyz = prefix_abc_xyz (instead of splitting it like abc_xyz = abc_prefix_xyz).

Again, I think this is done.

> 3. Consistenly use exactly the same signal names.
E.g. c_rdp = prefix_c_rdp (instead of tiny differences like c_rdp = prefix_c_rd, without the "p").

@silabs-robin are there any specific instances of inconsistent use of signal names left?

> 4. Keep a separation between "raw immediate" vs "interpreted immediate", etc [0]
E.g. blabla_raw and blabla_interpreted (because toggle coverage of {imm[20:1], 1'b 0} will never be reached)

@silabs-robin are there any specific instances of inconsistent use of signal names left?


So, I _think_ (1) and (2) are resolved and will defer to @silabs-robin for (3) and (4).

Comment (silabs-robin) [reply lib/uvm_agents/uvma_isacov/uvma_isacov_instr.sv]: Hi. I think it is good enough to merge now. Will approve.

**Review by silabs-robin — APPROVED** [review]

**Comment by AyoubJalali** [comment]
@MikeOpenHWGroup @silabs-robin, I added also a commit about CL-TYPE instruction imm decode (it wasn't implemented)

**Comment by AyoubJalali** [comment]
@MikeOpenHWGroup Could we merge ?

**Review by MikeOpenHWGroup — APPROVED** [review]

**Comment by MikeOpenHWGroup** [comment]
Hi @AyoubJalali, the updates you propose _look_ good to me, but I see that this PR is failing the CI check and that one of the failures is Compliance.  I do not understand how a change to ISACOV could cause a Compliance failure, but I recommend you have @ASintzoff review this before merging.

**Comment by ASintzoff** [comment]
> Hi @AyoubJalali, the updates you propose _look_ good to me, but I see that this PR is failing the CI check and that one of the failures is Compliance. I do not understand how a change to ISACOV could cause a Compliance failure, but I recommend you have @ASintzoff review this before merging.

The compliance failure was due to discrepancy between spike and cva6 misa values (A extension). It is fixed in [a CVA6 merged PR.](https://github.com/openhwgroup/cva6/pull/1323)

**Pull request merged**
